### DISCUSSION TITLE
Questions about `pgconn.consume_input()` and `pgconn.notifies()`

### DISCUSSION THREAD

**Discussion by vZ2JEZsZW** [psycopg/psycopg#672] [Q&A] [answered]
Hi.

1. Is it ok to use `pgconn.consume_input()` and `pgconn.notifies()` to get notifications?
2. Does calling `pgconn.notifies()` until it returns `None` clears accumulated notifications as `notifies.clear()` did?

```python
while True:
    with conn_lock:
        conn.pgconn.consume_input()
        ns = []
        while True:
            n = conn.pgconn.notifies()
            if n is None:
                break
            else:
                ns.append(n)

    if len(ns) > 0:
        # notify other threads there is work to handle
        pass 

    _ = select.select([conn], [], [])

```

**Comment by dvarrazzo** [comment, marked as answer]
The `pgconn.notifies()` is a low-level wrapper to the [`PQnotifies()`](https://www.postgresql.org/docs/current/libpq-notify.html) function.

> Is it ok to use pgconn.consume_input() and pgconn.notifies() to get notifications?

It is more low level than it should be. Ideally you should use the [`conn.notifies()`](https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.notifies) iterator. I know that there is an issue with stopping it (#340) and I would like to address it for the next 3.2 release.

`PGconn.notifies()` returns `pq.PGnotify` objects, which are rougher than `psycopg.Notify` (their attributes are bytes, not strings, and the attribute names are silly).

In general, everything I know about how to consume notifies is encapsulated in the chain of calls that starts from the `Connection.notify()` method. You should follow the path from that method below to figure out what to do if you want to consume them manually.


> Does calling pgconn.notifies() until it returns None clears accumulated notifications as notifies.clear() did?

`pgconn.notifies()` should clear the notification from the pgconn, yes. In psycopg2 the `notifies` attribute is just a normal Python queue that continuously receive notifications from the underlying PGconn.

**Reply by vZ2JEZsZW** [reply]
> You should follow the path from that method below to figure out what to do if you want to consume them manually.

I tried to read it but it's little bit complicated for me although it allowed me to find these two methods.

> PGconn.notifies() returns pq.PGnotify objects, which are rougher than psycopg.Notify (their attributes are bytes, not strings, and the attribute names are silly).

I see. But currently I don't look into the payloads.

I tried to use `conn.notifies()`. For some reason the waiting part of `notifies()` to a large extent precludes progress of other threads using the connection as if it intermittently holds some internal lock.

Another thing with `conn.notifies()` is that I can't figure out how to get just already accumulated notifications out of it without waiting for new ones to arrive and blocking. It blocks if I iterate past what is already there and I don't know the number in advance to not `next()` further. So it is great that it blocks, but in some cases I don't want it to.

The existing script I am updating to psycopg2 uses single connection for multiple threads with one them being that "listener" thread. And as I can't figure out how to use `conn.notifies()` for that particular case I just made it psycopg2-way. Probably `conn.notifies()` just warrants redesign my program in cleaner way which establish a separate connection for a thread with `conn.notifies()` and another for threads doing usual transactions.

**Reply by vZ2JEZsZW** [reply]
I wanted also to ask, when one uses `pgconn.consume_input()` with `pgconn.notifies()` or `conn.notifies()` does one retrieves all notifications sent by the server including those arrived while doing usual transaction work? For example, when one thread not doing that loop of `consume_input()` with `pgconn.notifies()`, other thread uses the connection and notifications may happen too.
Should I collect them separately with `add_notify_handler()`?

**Reply by vZ2JEZsZW** [reply]
slowing progress of other threads when `conn.notifies()` is used may actually be a result of some my bug though.

**Reply by dvarrazzo** [reply]
`pgconn.notifies()` is a low level and an internal function. It is not easy to use, and you are on your own. You need to know entirely how the BE-FE query protocol works. I don't think is a good idea. Use `conn.notifies()` or `conn.add_notify_handler()`.

Notifications usually don't work within transactions: they are sent when a transaction is committed.  Please read the [database documentation](https://www.postgresql.org/docs/current/sql-notify.html). Use two psql shells to run some experiments and check the interaction between notifications and transactions.

**Reply by vZ2JEZsZW** [reply]
I know that they are delivered in between transactions and shipped only on commit. The `conn.notifies()` combines waiting with reading notifications, while the select-recipe you offered in psycopg2 have them separate. The latter allowed to use single connection from several threads for both normal work and notifications thread. During idle periods listener returns from select, acquires a lock, reads `conn.notifies` and calls `conn.notifies.clear()`, then it releases the lock and signals other threads, which in turn acquire the lock and do transactions acting upon that notify. This also allowed any thread to use connection in arbitrary moments (say request or timer) as listener only holds it for reading `conn.notifies` property but not when being in `select()`. 
My questions were related to upgrading that existing program from psycpg2 to psycopg3 and seems that `conn.notifies()` was meant for different design, `pgconn.*` is low level internal interface, and it is not clear whether in absence of `add_notify_handler()` notifications discovered during normal statements processing are collected too for eventual retrieval with `conn.notifies()`. Now I realize it's better to keep that program psycpg2 for now.
Thank you for your responses.

**Reply by dvarrazzo** [reply]
In psycopg 3 we added `conn.notifies()` as high level generator but during its use it's not possible  to use the connection for other queries.

You can do something close to psycopg2 by using a notify handler and use a "select 1" instead of `conn.poll()` when you are told that there is a notification:

```python
>>> import psycopg
>>> conn = psycopg.connect("", autocommit=True)

>>> def notify_handler(n):
...     print(n)

>>> conn.add_notify_handler(notify_handler)

>>> conn.execute("listen foo");

>>> import select
>>> select.select([conn], [], [], 10)

# Meanwhile, in psql:
# =# notify foo, 'bar';
# NOTIFY

# `select()` returns as soon as the notification is sent
([<psycopg.Connection [IDLE] (database=piro) at 0x7fa094840c10>], [], [])

# Now, the cleanest thing to do is to use a `select 1` to trigger the consume input cycle
>>> conn.execute("select 1");

# this is printed by the notify_hander
Notify(channel='foo', payload='bar', pid=3977768)
```

I am thinking about extending `conn.notifies()` adding parameters `timeout` and `stop_after` to allow to stop after a certain timeout and after a certain number of notifications have been received. The implementation looks like the following for the moment, and I would like to add tests and check if it fits the bill to close #340.

```python
    def notifies(
        self, *, timeout: Optional[float] = None, stop_after: Optional[int] = None
    ) -> Generator[Notify, None, None]:
        """
        Yield `Notify` objects as soon as they are received from the database.

        :param timeout: maximum amount of time to wait for notifications.
            `!None` means no timeout.
        :param stop_after: stop after receiving this number of notifications.
            You might actually receive more than this number if more than one
            notifications arrives in the same packet.
        """
        nreceived = 0
        deadline = monotonic() + timeout if timeout is not None else None

        # Allow interrupting the wait with a signal
        timeout = 0.1

        while True:
            if deadline:
                timeout = min(0.1, deadline - monotonic())
                if timeout < 0.0:
                    break

            # Collect notifications. Also get the connection encoding if any
            # notification is received to makes sure that they are consistent.
            try:
                with self.lock:
                    ns = self.wait(notifies(self.pgconn), timeout=timeout)
                    if ns:
                        enc = pgconn_encoding(self.pgconn)
            except e._NO_TRACEBACK as ex:
                raise ex.with_traceback(None)

            # Emit the notifications received.
            for pgn in ns:
                n = Notify(pgn.relname.decode(enc), pgn.extra.decode(enc), pgn.be_pid)
                yield n
                nreceived += 1

            # Stop if we have received enough notifications.
            if stop_after is not None and nreceived >= stop_after:
                break
```

Would this interface work for your use case? Do you want to try the above? (as an external function, replacing `self` with `conn`).

**Reply by dvarrazzo** [reply]
The MR #673 proposes an interruptable notification receiver in order to receive only a certain amount of notifications or only for a certain time.

**Reply by vZ2JEZsZW** [reply]
I checked the function you provided running it in a separate script, I didn't yet tried to integrate it into mentioned program. 

In most cases it reacquires the lock itself not leaving it to others waiting. It can take e.g. 50 iterations * 0.1 = 5s before other thread acquires it, sometimes longer. Therefore, entering `conn.transaction()`, calling `execute()` and committing may take about 15 seconds.

It seems that as a workaround, one can add `time.sleep(0)` after leaving `with self.lock:`, this allows other waiters to acquire it, so it can takes like 0.3s for other thread to execute a transaction. 
Thanks for this recipe.

**Reply by vZ2JEZsZW** [reply]
I also think now that probably having a separate connection for listening thread is pretty justified by cleaner and simpler program design it results in. Also probably demanding `notifies()` of being capable of sharing a connection is too much, given than one have other options: use separate connection, recipe you just provided with `sleep(0)` added or even select() trick. Probably it is somehow specialized requirement after all, to share a connection for two different uses.

**Reply by vZ2JEZsZW** [reply]
I found that in case one wants preemptability by other threads, `timeout` parameter is useful in that one can call it as `notifies_new(conn, timeout=0.1, stop_after=None)`  thus limiting the inner loop to one iteration, which consequently allows to place `sleep(0)` outside of it, in user code. This works at least for my use case and replaces `select()` recipe.

As of `stop_after` I am not using it because now `timeout` is used in my case. But I think `stop_after=1` would be useful if one want to retrieve n>=1 already accumulated items without blocking and react to them by single action, otherwise there is no way to do it.

**Reply by dvarrazzo** [reply]
Hello,

> In most cases it reacquires the lock itself not leaving it to others waiting.

With the design proposed, `notifies()` still need to be used in a collaborative way together with other users of the connection. But yes, using a short timeout may be a good workaround.

Note that if your connection is pretty busy querying, you can use a notify callback, which will gather the notifications as a byproduct of the others queries' processing. The callback may be pushing the notifications received to a queue and your program can use them from there.

Please note that the jobs failed in #673 show that the wait functions, except the one distributed in `psycopg[binary]`, hang if no notification is received. It's something we need to investigate.
